supabase/functions/_script/backup_and_cleanup.ts

```diff
@@ -0,0 +1,111 @@
+import { createClient } from 'https://esm.sh/@supabase/supabase-js@^2.2.3';
+import 'https://deno.land/x/dotenv/load.ts';
+import { r2 } from '../_utils/r2.ts';
+import type { Database } from '../_utils/supabase.types.ts';
+import * as fs from 'https://deno.land/std/fs/mod.ts';
+import * as path from 'https://deno.land/std/path/mod.ts';
+
+const supabaseUrl = Deno.env.get('SUPABASE_URL');
+const supabaseServiceRoleKey = Deno.env.get('SUPABASE_SERVICE_ROLE_KEY');
+const backupFolder = 'backup';
+
+if (!supabaseUrl || !supabaseServiceRoleKey) {
+  console.error('Error: SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY environment variables are required.');
+  Deno.exit(1);
+}
+
+function useSupabase() {
+  const options = {
+    auth: {
+      autoRefreshToken: true,
+      persistSession: true,
+      detectSessionInUrl: false,
+    },
+  };
+  return createClient<Database>(supabaseUrl, supabaseServiceRoleKey, options);
+}
+
+async function backupAndDeleteTable(tableName: string) {
+  const currentDate = new Date();
+  const backupFilenames: string[] = [];
+  const pageSize = 1000;
+
+  try {
+    if (!await fs.exists(backupFolder)) {
+      await fs.ensureDir(backupFolder);
+    }
+
+    const supabase = useSupabase();
+    const daysToKeep = 30;
+    let offset = 0;
+    let hasMoreData = true;
+
+    const cutoffDate = new Date(currentDate.getTime() - daysToKeep * 24 * 60 * 60 * 1000); // Declare cutoffDate outside the loop
+
+    while (hasMoreData) {
+      const query = `
+        SELECT * FROM ${tableName}
+        WHERE updated_at < '${cutoffDate.toISOString()}'
```

Comment:
the only way this to not fail would be to set the timeout for the session, but not sure we can do it

Reply:
IG setting a session timeout is not feasible in this script. Proper error handling and retry mechanisms should suffice.

Reply:
@ayusham001 i understand your point but you have to understand the current table:
![CleanShot 2023-09-27 at 16 25 29@2x](https://github.com/Cap-go/capgo/assets/4084527/49a5dd18-4eed-4ccd-af07-3c22c6bc312f)
This request in loop will take too much CPU for too long that will break prod

Reply:
Thank you for information.
It will indeed put a lot of overhead, so should i put the timeout function right away or not?

Reply:
@rishi-raj-jain your suggestions on this?

Reply:
@ayusham001 why is adding a session timeout not possible in this file?

Reply:
You can just make something like:

when it's in while (hasMoreData) just add

await new Promise(r => setTimeout(r, 3000))

as the next line to process, that shall do it, correct?

Reply:
Yes, just asked for a suggestion
M on it.

Reply:
> Yes, just asked for a suggestion

Oh yeah that's fine, I just meant to ask simply no other intentions. Pardon if the framing seemed not apt.

Reply:
Also, appreciate the patience you have man, you've been editing this mutliple times now.

Reply:
No I totally get the concern.
Also have pushed the changes

Reply:
Thank you @rishi-raj-jain Grateful for the compliment, it's just part of the work.